### PR TITLE
Remove dependencies between Numba's Cpython and NumPy module

### PR DESCRIPTION
**Pull request by kc611**
As titled.

This PR intends to remove dependency of Cpython module on NumPy module and adds independent builtins that are seperate for NumPy and Cpython modules.

### CODE REVIEW

File: numba/np/math/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from numba.np.math import cmathimpl
+from numba.np.math import mathimpl
+from numba.np.math import numbers
```
Comment (sklam) [inline numba/np/math/__init__.py:3]: Seems unnecessary. Use-sites of these modules are already importing them explicitly.

File: numba/np/math/cmathimpl.py [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+import math
+
+from numba.core.imputils import Registry, impl_ret_untracked
+from numba.core import types, cgutils
+from numba.core.typing import signature
+from numba.cpython import builtins, mathimpl
+from numba.core.extending import overload
+
+registry = Registry('cmathimpl')
+lower = registry.lower
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:16]: Until the new types are implemented, the use of `lower` and `overload` are creating duplicated entries. Should all these be commented out for now?

File: numba/np/math/cmathimpl.py [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+"""
+
+
+import cmath
+import math
+
+from numba.core.imputils import Registry, impl_ret_untracked
+from numba.core import types, cgutils
+from numba.core.typing import signature
+from numba.cpython import builtins, mathimpl
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:12]: `mathimpl` should be from `from . import mathimpl`

File: numba/np/math/mathimpl.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+from numba.core.imputils import Registry, impl_ret_untracked
+from numba import typeof
+from numba.core import types, utils, config, cgutils
+from numba.core.extending import overload
+from numba.core.typing import signature
+from numba.cpython.unsafe.numbers import trailing_zeros
+
+
+registry = Registry('mathimpl')
+lower = registry.lower
```
Comment (sklam) [inline numba/np/math/mathimpl.py:22]: Same comment about `lower`. Should all `@lower` be commented out for now?

File: numba/np/math/numbers.py [hunk 5]
```diff
@@ -0,0 +1,1395 @@
...
+import numpy as np
+import operator
+
+from llvmlite import ir
+from llvmlite.ir import Constant
+
+from numba.core.imputils import (lower_builtin, lower_getattr,
+                                    lower_getattr_generic, lower_cast,
+                                    lower_constant, impl_ret_borrowed,
+                                    impl_ret_untracked)
```
Comment (sklam) [inline numba/np/math/numbers.py:13]: Same comment here. Should all registrations be commented out?

**Comment by sklam** [comment]
File history look ok under git blame

File: numba/cpython/rangeobj.py [hunk 6]
```diff
@@ -207,10 +245,9 @@ def  codegen(context, builder, sig, args):
             intp_t = context.get_value_type(types.intp)
             iterobj = context.make_helper(builder, iterty, value=value)
             arrayty = iterty.array_type
-            ary = make_array(arrayty)(context, builder, value=iterobj.array)
-            shape = cgutils.unpack_tuple(builder, ary.shape)
+            outer_shape = get_ary_shape_outermost(arrayty, iterobj.array, context, builder)
             # array iterates along the outer dimension
-            return impl_ret_untracked(context, builder, intp_t, shape[0])
+            return impl_ret_untracked(context, builder, intp_t, outer_shape)
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/cpython/rangeobj.py:250]: We can, but the function uses `ListIterInstance` imported from `numba.cpython` which will lead to a unwanted extra dependency of `cpython` in `core`.

File: numba/np/math/__init__.py (hunk 1 above)
Comment (kc611) [reply numba/np/math/__init__.py]: Fixed in [d17b0d8](https://github.com/numba/numba/pull/9437/commits/d17b0d8a3814d1b2b501366b83dcca2f1b75b3c6)

File: numba/np/math/mathimpl.py (hunk 4 above)
Comment (kc611) [reply numba/np/math/mathimpl.py]: Fixed in e91e8ca

File: numba/np/math/cmathimpl.py (hunk 3 above)
Comment (kc611) [reply numba/np/math/cmathimpl.py]: Fixed in e91e8ca

File: numba/np/math/numbers.py (hunk 5 above)
Comment (kc611) [reply numba/np/math/numbers.py]: Fixed in [93715b1](https://github.com/numba/numba/pull/9437/commits/93715b151e4350842fb227290eeda14d9f65c5ac)

File: numba/np/math/cmathimpl.py (hunk 2 above)
Comment (kc611) [reply numba/np/math/cmathimpl.py]: Fixed in [93715b1](https://github.com/numba/numba/pull/9437/commits/93715b151e4350842fb227290eeda14d9f65c5ac)

File: numba/cpython/rangeobj.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sklam) [reply numba/cpython/rangeobj.py:250]: The problem is that the current changes are copying/inlining ideas from `numpy` module into `cpython` module. 

Some passes needs to know about both `cpython` and `numpy`. I don't think we need to have a clean separation of `cpython` and `numpy` deps from `core`. At least, not needed for numpy2 type system update. 

In the future, to cleanly separate, `length_of_iterator` should have some `singledispatch` which `cpython` and `numpy` module can extend. 

For now, just move `length_of_iterator` to `inline_closurecall` should be sufficient and avoid code duplication.

[the remark this answers is not in the document]
Comment (kc611) [reply numba/cpython/rangeobj.py:250]: Also would be a good idea then to remove `get_ary_shape_outermost` function and use `make_array` as it did before ?

[the remark this answers is not in the document]
Comment (sklam) [reply numba/cpython/rangeobj.py:250]: correct

File: numba/np/math/mathimpl.py [hunk 7]
```diff
@@ -14,6 +14,7 @@
 from numba.core import types, config, cgutils
 from numba.core.typing import signature
 from numba.cpython.unsafe.numbers import trailing_zeros
+from numba.cpython.mathimpl import _unsigned
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/np/math/mathimpl.py]: Fixed the above and [comment](https://github.com/numba/numba/pull/9437#pullrequestreview-1909795274) in [399bbf6](https://github.com/numba/numba/pull/9437/commits/399bbf6c6152b15f70c3110f32d83959c94b997e)

File: numba/np/math/cmathimpl.py [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+# @lower(cmath.isfinite, types.Complex)
+def isfinite_float_impl(context, builder, sig, args):
+    [typ] = sig.args
+    [value] = args
+    z = context.make_complex(builder, typ, value=value)
+    res = is_finite(builder, z)
+    return impl_ret_untracked(context, builder, sig.return_type, res)
+
+
+@overload(cmath.rect)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:57]:
```suggestion
# @overload(cmath.rect)
```
it's duplicating registration with `cpython` https://github.com/numba/numba/blob/399bbf6c6152b15f70c3110f32d83959c94b997e/numba/cpython/cmathimpl.py#L57

File: numba/np/math/cmathimpl.py [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+    [z, base] = args
+
+    def log_base(z, base):
+        return cmath.log(z) / cmath.log(base)
+
+    res = context.compile_internal(builder, log_base, sig, args)
+    return impl_ret_untracked(context, builder, sig, res)
+
+
+@overload(cmath.log10)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:169]:
```suggestion
# @overload(cmath.log10)
```
duplication of https://github.com/numba/numba/blob/399bbf6c6152b15f70c3110f32d83959c94b997e/numba/cpython/cmathimpl.py#L169

File: numba/np/math/cmathimpl.py [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+        """cmath.log10(z)"""
+        z = cmath.log(z)
+        # This formula gives better results on +/-inf than cmath.log(z, 10)
+        # See http://bugs.python.org/issue22544
+        return complex(z.real / LN_10, z.imag / LN_10)
+
+    return log10_impl
+
+
+@overload(cmath.phase)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:186]:
```suggestion
# @overload(cmath.phase)
```

File: numba/np/math/cmathimpl.py [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+
+    if not isinstance(x, types.Complex):
+        return
+
+    def impl(x):
+        return math.atan2(x.imag, x.real)
+    return impl
+
+
+@overload(cmath.polar)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:198]:
```suggestion
# @overload(cmath.polar)
```

File: numba/np/math/cmathimpl.py [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+# @lower(cmath.cos, types.Complex)
+def cos_impl(context, builder, sig, args):
+    def cos_impl(z):
+        """cmath.cos(z) = cmath.cosh(z j)"""
+        return cmath.cosh(complex(-z.imag, z.real))
+
+    res = context.compile_internal(builder, cos_impl, sig, args)
+    return impl_ret_untracked(context, builder, sig, res)
+
+@overload(cmath.cosh)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:279]:
```suggestion
# @overload(cmath.cosh)
```

File: numba/np/math/cmathimpl.py [hunk 13]
```diff
@@ -0,0 +1,542 @@
...
+def sin_impl(context, builder, sig, args):
+    def sin_impl(z):
+        """cmath.sin(z) = -j * cmath.sinh(z j)"""
+        r = cmath.sinh(complex(-z.imag, z.real))
+        return complex(r.imag, -r.real)
+
+    res = context.compile_internal(builder, sin_impl, sig, args)
+    return impl_ret_untracked(context, builder, sig, res)
+
+@overload(cmath.sinh)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:319]:
```suggestion
# @overload(cmath.sinh)
```

File: numba/np/math/cmathimpl.py [hunk 14]
```diff
@@ -0,0 +1,542 @@
...
+    def tan_impl(z):
+        """cmath.tan(z) = -j * cmath.tanh(z j)"""
+        r = cmath.tanh(complex(-z.imag, z.real))
+        return complex(r.imag, -r.real)
+
+    res = context.compile_internal(builder, tan_impl, sig, args)
+    return impl_ret_untracked(context, builder, sig, res)
+
+
+@overload(cmath.tanh)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:357]:
```suggestion
# @overload(cmath.tanh)
```

File: numba/np/math/cmathimpl.py [hunk 15]
```diff
@@ -0,0 +1,542 @@
...
+            s1 = cmath.sqrt(complex(1. - z.real, -z.imag))
+            s2 = cmath.sqrt(complex(1. + z.real, z.imag))
+            real = 2. * math.atan2(s1.real, s2.real)
+            imag = math.asinh(s2.real * s1.imag - s2.imag * s1.real)
+            return complex(real, imag)
+
+    res = context.compile_internal(builder, acos_impl, sig, args)
+    return impl_ret_untracked(context, builder, sig, res)
+
+@overload(cmath.acosh)
```
Comment (sklam) [inline numba/np/math/cmathimpl.py:413]:
```suggestion
# @overload(cmath.acosh)
```

File: numba/np/math/numbers.py [hunk 16]
```diff
@@ -0,0 +1,1390 @@
...
+import numbers
+
+import numpy as np
+
+from llvmlite import ir
+from llvmlite.ir import Constant
+
+from numba.core.imputils import impl_ret_untracked
+from numba.core import typing, types, errors, cgutils
+from numba.cpython.unsafe.numbers import viewer
```
Comment (sklam) [inline numba/np/math/numbers.py:11]: This will require a `np` version. This can be done later in subsequent PRs.

**Review by sklam — APPROVED** [review]

**Pull request merged**